src/axom/core/docs/sphinx/core_containers.rst

```diff
@@ -137,6 +145,69 @@ The output of this example is::
   Range-based for loop over ArrayView c yields: 1 5 6 9 1 4
   Standard for loop over ArrayView c yields: 1 5 6 9 1 4
 
+########################
+Using Arrays in GPU Code
+########################
+
+Instead of writing kernels and device functions that operate on raw pointers, we can use ``ArrayView``
+in device code. The basic "workflow" for this process is as follows:
+
+  1. Create an ``Array`` allocated in device-accessible memory via either specifying an allocator ID
+     or using a class template parameter for the desired memory space.
+  2. Write a kernel that accepts an ``ArrayView`` parameter **by value**.
+  3. Create an ``ArrayView`` from the ``Array`` to call the function.  For non-templated kernels
+     an implicit conversion is provided.
+
+
+The full template signature for ``Array`` (``ArrayView`` has an analogous signature) is
+``Array<typename T, int DIM = 1, MemorySpace SPACE = MemorySpace::Dynamic>``.  Of particular interest
+is the last parameter, which specifies the memory space in which the array's data are allocated.
+The default, ``Dynamic``, means that the memory space is set via an allocator ID at runtime.
+
+.. note:: Allocating ``Array``s in different memory spaces is only possible when Umpire is available.
+
+Setting the ``MemorySpace`` to an option other than ``Dynamic`` - e.g., ``MemorySpace::Device`` provides
+a compile-time guarantee that data can always be accessed from a GPU.  "Locking down" the memory space at
+compile time can help to prevent illegal memory accesses and segmentation faults when pointers are dereferenced
+from the wrong execution space.
+
+For example, a GPU kernel can require that its argument arrays be allocated in a specific memory space.
+
+.. literalinclude:: ../../examples/core_containers.cpp
+   :start-after: _cuda_kernel_start
+   :end-before: _cuda_kernel_end
+   :language: C++
+
+To illustrate how different memory spaces can be required, the inputs happen to be in unified memory and the output in
+device memory.
+
+The following snippet illustrates how one would create and initialize the inputs/outputs to this kernel.  Note
+how a "Dynamic" ``Array`` can be converted into an ``Array`` whose memory space is locked down.
+
+.. literalinclude:: ../../examples/core_containers.cpp
+   :start-after: _cuda_array_create_start
+   :end-before: _cuda_array_create_end
+   :language: C++
+
```

Comment:
What happens if the user does something like specify the memory space with a template parameter but provide an incompatible allocator?  (A compile-time failure would be awesome!)  I can see this happening through code modification or incremental development/debugging.

Reply:
In this case the incompatible allocator is ignored:
https://github.com/LLNL/axom/blob/6644fd6526506489e9f18962253875ba5af1d696/src/axom/core/Array.hpp#L647-L649

That being said, I still have a warning in this case:
https://github.com/LLNL/axom/blob/6644fd6526506489e9f18962253875ba5af1d696/src/axom/core/Array.hpp#L651-L661

Reply:
Thanks for clarifying.  Just as a suggestion, consider adding a sentence like

```suggestion
.. note:: Unless the Dynamic memory space is in use, the ``Array`` constructor will
          ignore an allocator ID that doesn't match its memory space, and in debug
          builds will print a warning at runtime.
```